Ies testing

joakim-hove opened the pull request:
The important point of this PR is to refactor the `ies_enkf` code to wrangle out a function `ies_enkf_initX` which can be called as an alternative to the normal `std_enkf_initX`. This PR demonstrates that the two calls are equivalent with a test - a subsequent PR will change to actually use the `ies_enkf` implementation as the only initX implementation.

ertomatic commented:
Can one of the admins verify this patch?

joakim-hove commented:
This is ready for review now.

joakim-hove commented:
Ohhh my - it is hard to get all of these tests through at the same time ...

markusdregi commented:
> Ohhh my - it is hard to get all of these tests through at the same time ...

Notice that for rerunning tests you can click `details` on the failing job, followed by the `rerun` button ;)

We have seen some flaky tests lately, but believe that the situation has improved vastly. If you still see flaky tests it would be very useful if you could check existing issues to see whether the flaky test is already documented in an issue and:
- if it is, verify that you have also hit this flakiness by a small comment in the issue,
- it is is not apparent in an issue, create a new issue with a copy of the error and a URL to the job :)

joakim-hove commented:
> Notice that for rerunning tests you can click details on the failing job, followed by the rerun button ;)

Thank you! It actually pulled through now, but I will update an issue according to the instruction if I get problems again.

dafeda commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -330,37 +233,128 @@ void ies_enkf_updateA(
+    }
+    costf = costf / ens_size;
+
+    teclog(W, D, DW, "iesteclog.dat", ens_size, iteration_nr, nrobs_inp);
```
Not really relevant, but what is a `tec`?

joakim-hove replied:
This is call to produce output for the [Tecplot](https://www.tecplot.com/) application. I wanted to remove it - but I am to polite a guy ....

dafeda replied:
First time I hear of Tecplot, thanks!
If it's not used we should definitely remove it, no need to be polite about it :)

dafeda replied:
I'll ask around a bit.

joakim-hove replied:
> If it's not used 

@geirev used to enjoy using it.

dafeda replied:
Looks like we can get rid of Tecplot.
What do you think, should we remove it in this PR or should I create an issue and remove it at a later stage?

joakim-hove replied:
If this can be merged as is that would be good, then I Will make a new PR removing tecplot afterwards.

If you have more on this I Will update tomorrow.

dafeda replied:
Sure!

I'll try to get through this before dinner, but it's probably a good idea to have another pairs of eyes on this.

dafeda commented on libres/tests/analysis/test_compare_initX.cpp:
```diff
@@ -0,0 +1,181 @@
+        obs_block_iset(ob, iobs, true_model.eval(x), 0.25);
+    }
+
+    {
```
Nitpick sorry - why do we need the brace here?

joakim-hove replied:
Looks cool!

dafeda replied:
Hehe no wonder you want to up the limit to 120 lines! :D

eivindjahren replied:
I think for the tests, replacing scope brackets with [BDD-style sections](https://github.com/catchorg/Catch2/blob/devel/examples/120-Bdd-ScenarioGivenWhenThen.cpp) would improve readability, and give more informative output when things fail. In this case I suggest:
```cpp
GIVEN("IES calculation matrices"){
        matrix_type *S = meas_data_allocS(meas_data);
        matrix_type *R = obs_data_allocR(obs_data);
        matrix_type *E = obs_data_allocE(obs_data, rng, ens_size);
        matrix_type *D = obs_data_allocD(obs_data, E, S);
        WHEN("X matrix is given the default settings") {
            std_enkf_data_type *std_enkf_data =
                static_cast<std_enkf_data_type *>(std_enkf_data_alloc());

            matrix_type *X1 = matrix_alloc(ens_size, ens_size);
            std_enkf_initX(std_enkf_data, X1, nullptr, S, R, nullptr, E, D,
                           rng);
            THEN("X matrix is equal to [...]") {
                matrix_type *X2 = matrix_alloc(ens_size, ens_size);
                ies_enkf_initX(std_enkf_get_truncation(std_enkf_data),
                           std_enkf_get_subspace_dimension(std_enkf_data),
                           std_enkf_data_get_inversion(std_enkf_data), S, R, E,
                           D, X2);

                REQUIRE(matrix_similar(X1, X2, 1e-10));
                std_enkf_data_free(std_enkf_data);
                matrix_free(X2);
            }
            matrix_free(X1);
        }
    }
}
```

joakim-hove replied:
Splitting a leaf node does really give anything - I'll drop the inner `THEN()`

dafeda commented on libres/tests/analysis/test_compare_initX.cpp:
```diff
@@ -0,0 +1,181 @@
+        ens.push_back(model(rng));
+
+    meas_block_type *mb = meas_data_add_block(meas_data, obs_key, 1, obs_size);
+    obs_block_type *ob =
```
What is `ob` used for?
I see that we set some values on line 55, but I don't understand how it's used.

joakim-hove replied:
Well - we need that handle to the interior of the obs_data in order to det the observer values

dafeda replied:
Ah, so doing `obs_block_iset(ob, iobs, true_model.eval(x), 0.25);` updates `obs_data` which is later used to initialize matrices?
My Python-brain hurts.

joakim-hove replied:
Yes that is correct; you should go to an old School C shrink!

dafeda replied:
Most definitely!

dafeda commented on libres/lib/analysis/ies/ies_enkf_config.cpp:
```diff
@@ -207,3 +210,18 @@ void ies_enkf_config_set_ies_logfile(ies_enkf_config_type *config,
+    double ies_max_step = ies_enkf_config_get_ies_max_steplength(ies_config);
+    double ies_min_step = ies_enkf_config_get_ies_min_steplength(ies_config);
+    double ies_decline_step =
+        std::max(ies_enkf_config_get_ies_dec_steplength(ies_config), 1.1);
```
Why 1.1?
A comment would be nice I think.

joakim-hove replied:
Have refactored a bit and added comment.

dafeda commented on libres/lib/analysis/ies/ies_enkf_config.cpp:
```diff
@@ -207,3 +210,18 @@ void ies_enkf_config_set_ies_logfile(ies_enkf_config_type *config,
+    double ies_decline_step =
+        std::max(ies_enkf_config_get_ies_dec_steplength(ies_config), 1.1);
+
+    double ies_steplength =
```
Is this a standard way to calculate step length?

joakim-hove replied:
Probably - have added a reference.

dafeda commented:
What do you think about adding a namespace called `ies` or something so that function don't have to start with `ies_enkf_`?

@eivindjahren did that [here](https://github.com/equinor/ert/blob/ca56ca4f03ca764993c14efd14f9a6a62f7d2e0a/libres/lib/analysis/update.cpp#L17), and I find it a nice improvement.

We can of course do that in another PR as well.

Would you mind having a look at this PR @eivindjahren ?

joakim-hove commented:
>What do you think about adding a namespace called ies or something so that function don't have to start with ies_enkf_?

I think that was a very good idea. I fear there might be some C linkage challenges - I Will look into it.

dafeda commented on libres/lib/analysis/ies/ies_enkf_data.cpp:
```diff
@@ -248,18 +248,14 @@ void ies_enkf_data_store_initialA(ies_enkf_data_type *data,
+void ies_enkf_data_allocateW(ies_enkf_data_type *data) {
     if (!data->W) {
         // We initialize data-W which will store W for use in next iteration                    (Line 9)
         bool dbg = ies_enkf_config_get_ies_debug(data->config);
```
`dbg` no longer in use.

dafeda commented on libres/lib/analysis/ies/ies_enkf_data.cpp:
```diff
@@ -248,18 +248,14 @@ void ies_enkf_data_store_initialA(ies_enkf_data_type *data,
         // We initialize data-W which will store W for use in next iteration                    (Line 9)
         bool dbg = ies_enkf_config_get_ies_debug(data->config);
+        int ens_size = bool_vector_size(data->ens_mask);
         int m_ens_size = std::min(ens_size - 1, 16);
```
Not related to this PR, but why 16?

dafeda replied:
`m_ens_size` is not in use though, so 16 doesn't matter :)

joakim-hove replied:
> Not related to this PR, but why 16?

It is a variety of 80 characters on a line ... purely visual formatting.

eivindjahren replied:
I think the important part here is that this is an unused variable so perhaps have another look at where this should go or whether it should be simply removed.

joakim-hove replied:
Everything removed

dafeda commented on libres/lib/analysis/ies/ies_enkf_data.cpp:
```diff
@@ -248,18 +248,14 @@ void ies_enkf_data_store_initialA(ies_enkf_data_type *data,
-void ies_enkf_data_allocateW(ies_enkf_data_type *data, int ens_size) {
+void ies_enkf_data_allocateW(ies_enkf_data_type *data) {
     if (!data->W) {
         // We initialize data-W which will store W for use in next iteration                    (Line 9)
```
Tiny typo, should be `data->W`.

dafeda commented on libres/lib/include/ert/analysis/ies/ies_enkf.hpp:
```diff
@@ -23,13 +23,30 @@
+#include <ert/analysis/ies/ies_enkf_data.hpp>
+
+#ifdef __cplusplus
+extern "C" {
```
Is this necessary now that we are using C++?
My understanding is that this is to prevent name-mangling and has to be used if we want to call the functions via `cwrap`.

joakim-hove replied:
My previous experience is that removing these very quickly ripple out of control - but I agree it is a good goal and I will try to remove them.

dafeda commented on libres/tests/analysis/ies/test_ies_enkf_main.cpp:
```diff
@@ -46,43 +43,135 @@ TEST_CASE("ies_enkf_linalg_extract_active", "[analysis]") {
-    matrix_type *E = matrix_alloc(state_size, ens_size);
-    ies_enkf_linalg_extract_active(data, E, stdout, false);
-    REQUIRE(matrix_equal(Ein, E));
+    {
```
So we need to talk about these braces! 😄 
I've heard about old-school variable-scoping in C, but is this necessary?

joakim-hove replied:
> So we need to talk about these braces! smile

We'll find a good counsellor ... :-)

> I've heard about old-school variable-scoping in C, but is this necessary?

It is not necessary; and I am not taking the fight. But quite frankly I find it a bit amusing that everyone dislikes these scopes so much - in my opinion the use of scopes like is the absolute lowest level of creating small manageable units - not a full function/method - but more than just flat code. For me this structure:
```C++
{
   int local_var = ...;
   local_var = 100;
   ...
}
..
{
    int local_var = 200;
    ...
}
```
helps with two things:
1. It creates a very clear visual grouping of related statements.
2. It guarantees that both blocks have a pristine version of `local_var` - with no pollution between the blocks.

In some cases the correct thing would probably be to extract functions, but this is a poor mans extracted function.

Anyway - I'll fold any day.

eivindjahren replied:
So these groups are quite uncommon. They come from a time where variables could only be declared at the beginning of a block, so it was a different time...

The norm for adding these blocks is that they would be there for a reason. So I often loose some time trying to figure out whether the additional curlies are doing anything to the code, before I figure out that they are just redundant. For instance, in C, a destructor for an object might be called early if you have inserted brackets. So its a bit of a surprise when they do nothing.

dafeda replied:
Thanks for the examples.

Having mostly used languages without curly braces, I find they pollute more than they clear up.
I'm sure this is just a matter of habit so not really important.

As for your second example where `local_var` is re-initialised, to me this signals that we perhaps want a function instead.
But also here it's a matter of taste, so "smaken er som baken" and all that :)

Since most people, or at least most of the people I've discussed extra curly braces with don't like them, let's do the democratic thing and leave them behind.

joakim-hove replied:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-scope: 

**Alternative formulation:** Don’t declare a name in an unnecessarily large scope.

dafeda replied:
I fully agree with the guidelines, and as far as I can tell they recommend splitting into several functions instead of creating scopes with curly braces.
Am I perhaps misunderstanding?

joakim-hove replied:
Well - the message is "Don't declare a name in an unnecessarily large scope."; that can be achieved in several ways. Splitting into functions is certainly a way to achieve that, but it can also be achieved by defining variables in a lexical scope with `{ ... }`.

Anyway - I promise on the grave om childhood dog that this is the last nag I will make on the topic.

eivindjahren commented:
> What do you think about adding a namespace called `ies` or something so that function don't have to start with `ies_enkf_`?
> 
> @eivindjahren did that [here](https://github.com/equinor/ert/blob/ca56ca4f03ca764993c14efd14f9a6a62f7d2e0a/libres/lib/analysis/update.cpp#L17), and I find it a nice improvement.
> 
> We can of course do that in another PR as well.
> 
> Would you mind having a look at this PR @eivindjahren ?

Having a look at it now. If the code is not in python then we can use cpp namespace instead of a c pseudo-namespace.

eivindjahren commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -372,46 +366,58 @@ void ies_enkf_updateA(
-                if (bool_vector_iget(ens_mask, iens)) {
-                    i++;
-                    matrix_iset_safe(E, m, i, matrix_iget(dataE, iobs, iens));
+static matrix_type *alloc_active(const matrix_type *full_matrix,
```
Could you also change the documentation for this function, it seems oddly specific to ies now, for what the function actually does.

joakim-hove replied:
Well it is `static` - of course it is specific to the `ies` implementation.

eivindjahren replied:
Sorry, I was ambigous, I was thinking the documentation above seems specific to ies, while the function is much more general.

joakim-hove replied:
OK - the function in question is an implementation detail for the three ies specific functions:
```C++
matrix_type *ies_enkf_alloc_activeE();
matrix_type *ies_enkf_alloc_activeW();
matrix_type *ies_enkf_alloc_activeA();
```
I'll shuffle things around so it hopefully becomes clearer.

eivindjahren commented on libres/tests/analysis/ies/test_ies_enkf_main.cpp:
```diff
@@ -46,43 +43,135 @@ TEST_CASE("ies_enkf_linalg_extract_active", "[analysis]") {
+    ies_enkf_data_update_ens_mask(data, ens_mask);
+    {
+        matrix_type *A = ies_enkf_alloc_activeA(data);
+        for (int i = 0; i < state_size; i++) {
```
What this loop does is very complicated, is there any way of making it simpler? Personally would prefer:
```
for (int i = 0; i < state_size; i++) {
            for (int j = 0; j < ens_size - 1; j++) {
                int j0 = j;
                if(j > 0) {
                   j0++;
                }
                REQUIRE(matrix_iget(A, i, j) == matrix_iget(A0, i, j0));
            }
        }
```

eivindjahren commented on libres/tests/analysis/ies/test_ies_enkf_main.cpp:
```diff
@@ -46,43 +43,135 @@ TEST_CASE("ies_enkf_linalg_extract_active", "[analysis]") {
+    }
+    ies_enkf_data_store_initialA(data, A0);
+
+    {
```
Instead of using a plain scope block here I would suggest using some more sections. The additional text given to the sections is also very helpful to the reader as tests often include arbitrary decisions so it can be hard to follow what is being attempted.

```cpp
SCENARIO("extracting active A matrix", "[analysis]") {
    GIVEN("Initial setup for ies update algorithm") {
        const int ens_size = 4;
        const int obs_size = 10;
        const int state_size = 10;
        ies_enkf_data_type *data = (ies_enkf_data_type *)ies_enkf_data_alloc();
        bool_vector_type *ens_mask = bool_vector_alloc(ens_size, true);
        bool_vector_type *obs_mask = bool_vector_alloc(obs_size, true);

        ies_enkf_config_set_ies_logfile(ies_enkf_data_get_config(data), "log");
        ies_enkf_data_open_log(data);
        ies_enkf_init_update(data, ens_mask, obs_mask, nullptr, nullptr, nullptr,
                             nullptr, nullptr, nullptr);
        ies_enkf_data_update_ens_mask(data, ens_mask);

        AND_GIVEN("inital A matrix (A0) with cells set to 1,2,3.."){
            matrix_type *A0 = matrix_alloc(state_size, ens_size);
            for (int i = 0; i < state_size; i++) {
                for (int j = 0; j < ens_size; j++)
                    matrix_iset(A0, i, j, i * ens_size + j);
            }
            ies_enkf_data_store_initialA(data, A0);

            THEN("extracting active A results in A0"){
                matrix_type *A = ies_enkf_alloc_activeA(data);
                REQUIRE(matrix_equal(A, A0));
                matrix_free(A);
            }

            WHEN("ensemble at index 1 is deactivated"){
                bool_vector_iset(ens_mask, 1, false);
                ies_enkf_data_update_ens_mask(data, ens_mask);
                THEN("Extracting A results in A0 minus index 1"){
                    matrix_type *A = ies_enkf_alloc_activeA(data);
                    for (int i = 0; i < state_size; i++) {
                        int i0 = i;
                        for (int j = 0; j < ens_size - 1; j++) {
                            int j0 = j + (j > 0);
                            REQUIRE(matrix_iget(A, i, j) == matrix_iget(A0, i0, j0));
                        }
                    }
                    matrix_free(A);
                }
            }

            matrix_free(A0);
        }
        bool_vector_free(ens_mask);
        bool_vector_free(obs_mask);
        ies_enkf_data_free(data);
    }
}
```

joakim-hove replied:
You'll get some sections

eivindjahren commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -93,182 +83,103 @@ static void printf_mask(FILE *log_fp, const char *name,
+static void log_active(FILE *log_fp, int ens_size, int ens_size_msk,
+                       int nrobs_input, int nrobs_msk, int state_size,
+                       const ies_enkf_data_type &data) {
+    fprintf(log_fp, "----active ens_size  = %d, total ens_size_msk = %d\n",
```
We have a logging utility now which should channel logs to the correct place. Is there a way of replacing these `fprintf`s with that now @dotfloat.

dotfloat replied:
```cpp
#include <ert/python.hpp>

namespace {
auto logger = ert::get_logger("ies");
} // namespace

[...]

logger->info("----active ens_size = {}, total end_size_msk = {}", ens_size, ens_size_msk);
```

eivindjahren commented on libres/lib/include/ert/analysis/ies/ies_enkf.hpp:
```diff
@@ -23,13 +23,30 @@
+                                    const matrix_type *W0);
+
+matrix_type *ies_enkf_alloc_activeE(const ies_enkf_data_type *data);
+matrix_type *ies_enkf_alloc_activeW(const ies_enkf_data_type *data);
```
Generally I would prefer if the `.hpp` included only only the external interface of `ies_enkf` to the remaining program. For testing purposes the prototype can be repeated in the corresponding cpp file. However, this is a design choice we have not discussed yet. Should we add a discussion for this @dotfloat ? Or do you have a better suggestion?

eivindjahren commented:
> > What do you think about adding a namespace called ies or something so that function don't have to start with ies_enkf_?
> 
> I think that was a very good idea. I fear there might be some C linkage challenges - I Will look into it.

I haven't found any usage of `ies_enkf` in `res/` so I second that we attempt this. I think it would greatly improve readability.

eivindjahren commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -490,12 +446,13 @@ void ies_enkf_linalg_extract_active_W(const ies_enkf_data_type *data,
     int m_ens_size = std::min(ens_size - 1, 16);
     int m_nrobs = std::min(nrobs - 1, 7);
     double nsc = 1.0 / sqrt(ens_size - 1.0);
+    double rcond;
```
This seems to remain a bit of a unused vestige now. Generally the old c programming style of declaring variables at the top of the scope is generally considered bad practice. Because it is unused (except to give matrix_dgesvx an output), its very confusing. So I suggest this variable is moved to where it is used:
```
double rcond;
matrix_dgesvx(Omega, YT, &rcond);
```
I think maybe `matrix_dgesvx` should accept `nullptr` to mean "Do not give me that output", which is a common practice. I am not sure whether lapack does that already, but I believe the underlying fortran code makes no such check.

joakim-hove replied:
> I think maybe matrix_dgesvx should accept nullptr to mean "Do not give me that output"

That is a good idea.

> [...] but I believe the underlying fortran code makes no such check.

Fortran is call by reference and makes no such check.

joakim-hove commented:
OK - I now have addressed most of the review comments, the changes are extensive:

1. I have added a namespace `ies::` - that was a very good idea. I have opted to use an explicit `ies::` in the implementation to avoid confusion with very similar unqualified symbols in the code.
2. Have removed additional logging / debug calls compared to the first version.
3. Have introduced BDD style `SCENARIO()` structure on the tests - although not as deeply nested as suggested by @eivindjahren 
4. The `{ ... }` which were pointed out in review have been refactored away, there might be others still remaining. I will remove them if pointed out.

eivindjahren commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -293,33 +157,25 @@ void ies_enkf_updateA(
      * ies_inversion=IES_INVERSION_SUBSPACE_EE_R (Nm^2).
      *
-     * See the enum: ies_inverson in ies_enkf_config.hpp:
+     * See the enum: ies_inverson in ies::enkf_config.hpp:
```
I think a search/replace went awry here.

joakim-hove replied:
yes

eivindjahren commented on libres/tests/analysis/ies/test_ies_enkf_main.cpp:
```diff
@@ -38,51 +35,142 @@ TEST_CASE("ies_enkf_linalg_extract_active", "[analysis]") {
+        for (int j = 0; j < ens_size; j++)
+            matrix_iset(W0, i, j, i * ens_size + j);
+    }
+    ies::enkf_linalg_store_active_W(data, W0);
```
Since these are being renamed already, there are a couple more prefixes to these names that I am wondering whether is best omitted. That is `ies::enkf_*`, `ies::linalg_*` and `ies::enkf_linalg_*`. It might be room for a `ies::linalg::` namespace, but it seems that all functions currently with `linalg` in their names are not purely about linear algebra.

joakim-hove replied:
I can do renames, but I do not fully understand what you request here. 

Must admit I feel this PR has been in the works quite long now and would be happy if it could be merged.

dafeda replied:
I'm sure future PR's will get merged much faster, as we are using this PR to discuss and agree on how we want the code to look, which is quite time-consuming.

eivindjahren replied:
Maybe you could split the rename into a different PR, then we can speed up the review of the other stuff and look at the renaming later.

joakim-hove replied:
Of course I can submit renames in a separate PR - that was implicitly what I suggested in me previous comment.

dafeda replied:
[Explicit is better than implicit](https://zen-of-python.info/explicit-is-better-than-implicit.html#2) 😄

eivindjahren replied:
Sorry, I was being ambiguous again, I was thinking that renaming and introduction of namespace belonged together so that 979d063 should go in the split out PR.

joakim-hove replied:
Sorry - still not entirely clear. Are now asking me to relegate the new `ies::` namespace to a new PR? Of course I can do that - but this review process feels quite schizophrenic from my end.

eivindjahren replied:
Sorry about the schizophrenic behavior, but I think its the best course of action. Then we will have discharged that point and can fast track the other parts :)

eivindjahren replied:
We should probably have considered making a separate PR for the ies namespace in the first place, but at least now we know :)

dafeda replied:
> I can do renames, but I do not fully understand what you request here.
> 
> Must admit I feel this PR has been in the works quite long now and would be happy if it could be merged.

We did not start looking at it before it was marked as "Ready for review", which was a couple of days ago if I remember correctly.
This perhaps add to the feeling that it's taking long to review.

joakim-hove commented:
I consider the current test failure a fluke - at least #2602 which is on top of this - runs green. I can not for myself see the test failure.

eivindjahren commented on libres/tests/res_util/test_matrix.cpp:
```diff
@@ -68,3 +69,28 @@ TEST_CASE("matrix_subtract_row_mean", "[res_util]") {
+            matrix_dgesvx(A, bx, nullptr);
+        }
+
+        for (int i = 0; i < N; i++)
```
For readability of the output in catch, it would be helpful to have a `THEN` block around this loop describing the expected behavior.

joakim-hove replied:
Well - if the test is reorganized as 
```
WHEN() { ... }

WHEN() { ... }

THEN() { ... }
```
it will fail - because only one section at a time is walked all the way to the leaf nodes, and with `THEN` in a separate section we will get two setup-and-calculate sections and one separate check section.

eivindjahren replied:
That is true, but you can use `GENERATE` to drop one of your `WHEN` blocks. Personally I would use:
```
SCENARIO("Solving a linear equation with dgesvx") {
    GIVEN("A diagonal matrix A") {
        std::size_t N = 4;
        matrix_type *A = matrix_alloc(N, N);
        for (int i = 0; i < N; i++)
            matrix_iset(A, i, i, 1.0 / (i + 1));

        AND_GIVEN("A vector B with all entries equal to 1"){
            matrix_type *bx = matrix_alloc(N, 1);
            for (int i = 0; i < N; i++)
                matrix_iset(bx, i, 0, 1);

            WHEN("Using dgesvx to solve the linear equation Ax=B") {
                double rcond;
                matrix_dgesvx(A, bx, GENERAGE(&rcond, nullptr));

                THEN("B is modified to be the solution") {
                    for (int i = 0; i < N; i++)
                        REQUIRE(matrix_iget(bx, i, 0) == (i + 1));
                }
            }
        }
    }
}
```

eivindjahren replied:
That is true, but you can avoid needing multiple `WHEN` blocks with a `GENERATE`. Personally I would use:
```

SCENARIO("Solving a linear equation with dgesvx") {
    GIVEN("A diagonal matrix A") {
        std::size_t N = 4;
        matrix_type *A = matrix_alloc(N, N);
        for (int i = 0; i < N; i++)
            matrix_iset(A, i, i, 1.0 / (i + 1));

        AND_GIVEN("A vector B with all entries equal to 1"){
            matrix_type *bx = matrix_alloc(N, 1);
            for (int i = 0; i < N; i++)
                matrix_iset(bx, i, 0, 1);

            WHEN("Using dgesvx to solve the linear equation Ax=B") {
                double * rcond = GENERATE((double*)malloc(sizeof(double)), NULL);
                matrix_dgesvx(A, bx, rcond);

                THEN("B is modified to be the solution") {
                    for (int i = 0; i < N; i++)
                        REQUIRE(matrix_iget(bx, i, 0) == (i + 1));
                }
                free(rcond);
            }
        }
    }
}
```

joakim-hove replied:
Well - this is different test; it only runs (and verifies) the matrix solve for `rcond != nullptr`.

eivindjahren replied:
I think it will run with `rcond=NULL` as well because of `GENERATE` :)

joakim-hove replied:
Ok - thanks; I had not understood GENERATE

joakim-hove replied:
OK - updated; valuable to learn about GENERATE.

eivindjahren reviewed (commented):
The IES refactor commit is a bit long so it would be helpful to have a commit message body describing what is done. Also, we prefer that commit message subjects are imperative (see the style guide in CONTRIBUTING.md)

eivindjahren commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -539,90 +434,62 @@ void ies_linalg_solve_S(const matrix_type *W0, const matrix_type *Y,
+            1.0 /
+                ((ens_size - 1) *
+                 (ens_size -
+                  1))); // since enkf_linalg_lowrankCinv solves (SS' + (N-1) R)^{-1}
```
The formatting of this line will probably be better if the comment on the end of the line is moved to 
on top of the line. Also, that comment is quite terse so it could benefit from being expanded upon if possible.

eivindjahren commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -539,90 +434,62 @@ void ies_linalg_solve_S(const matrix_type *W0, const matrix_type *Y,
-        matrix_pretty_fprint_submat(X3, "X3", "%11.5f", log_fp, 0, m_nrobs, 0,
-                                    m_ens_size);
+    {
+        /*    X3 = X1 * diag(eig) * X1' * H (Similar to Eq. 14.31, Evensen (2007))                                  */
```
This comment wraps on github so it would be great if it were shorter. Could probably just remove some redundant space before X3.

joakim-hove replied:
OK

eivindjahren commented on libres/lib/analysis/ies/ies_enkf_config.cpp:
```diff
@@ -75,7 +77,6 @@ ies_enkf_config_type *ies_enkf_config_alloc() {
     ies_enkf_config_set_ies_subspace(config, DEFAULT_IES_SUBSPACE);
     ies_enkf_config_set_ies_inversion(config, DEFAULT_IES_INVERSION);
     ies_enkf_config_set_ies_logfile(config, DEFAULT_IES_LOGFILE);
-    ies_enkf_config_set_ies_debug(config, DEFAULT_IES_DEBUG);
```
This would be a breaking change in the input config file,  so I don't think introducing breaking changes for the christmas release will be well received, if it impacts any of our users. Maybe we should introduce logging or warnings before making this change. Any input @oyvindeide?

joakim-hove replied:
The entire `bool dbg;` is removed - it was a pure development aid.

oyvindeide replied:
Agree that we should introduce a warning here, with a version for when it will be removed (in the near future). We have previously removed a few keywords, but the result of that has been that cases have printet out that it is not recognised. Adding a warning will make it consistent at least.

joakim-hove replied:
OK - I have done as I have suggested; if the user tries to invoke the `IES_DEBUG` subsetting a warning will be issued.

joakim-hove commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -539,90 +434,62 @@ void ies_linalg_solve_S(const matrix_type *W0, const matrix_type *Y,
-            nrsing += 1;
+        matrix_type *scaledR = matrix_alloc_copy(R);
+        matrix_scale(scaledR, nsc * nsc);
+        // std_enkf.cpp: 165
```
Removed

joakim-hove commented:
```ERROR: Unable to find a working compiler```

eivindjahren commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -883,14 +729,14 @@ bool ies_enkf_has_var(const void *arg, const char *var_name) {
             return true;
         else if (strcmp(var_name, ENKF_SUBSPACE_DIMENSION_KEY) == 0)
             return true;
+        else if (strcmp(var_name, IES_DEBUG_KEY) == 0)
```
This `else if` seem to be unnecessarily moved. Unless there is a reason to we should probably avoid it to not create a confusing diff.

joakim-hove replied:
fixed

eivindjahren commented on libres/tests/analysis/test_compare_initX.cpp:
```diff
@@ -0,0 +1,181 @@
+
+            matrix_type *X1 = matrix_alloc(ens_size, ens_size);
+            matrix_type *X2 = matrix_alloc(ens_size, ens_size);
+            std_enkf_set_string(std_enkf_data, INVERSION_KEY,
```
Should probably use `GENERATE` here as well: `std_enkf_set_string(std_enkf_data, INVERSION_KEY, GENERATE(STRING_INVERSION_SUBSPACE_RE, STRING_INVERSION_SUBSPACE_EXACT_R, STRING_INVERSION_SUBSPACE_EE_R))`, unless there is some difference that I am not catching. You can also use it for `std_enkf_set_double(std_enkf_data, ENKF_TRUNCATION_KEY_, GENERATE(0.90, 0.95, 0.75,...));` or if necessary
```cpp
if(GENERATE(true, false)){
    std_enkf_set_double(std_enkf_data, 0.9);
}
```

eivindjahren commented on libres/tests/analysis/test_compare_initX.cpp:
```diff
@@ -0,0 +1,181 @@
+
+    model(double a, double p) : amplitude(a), phase(p) {}
+
+    model(rng_type *rng) {
```
This model seems nice, and is probably going to see a lot more useage. The test only uses the first constructor though, and I think the second constructor is likely going to be superseded by the following useage pattern:
```cpp
double amplitude = GENERATE(take(10, random(1, 0.5))); // 10 different amplitudes
double phase = GENERATE(take(10, random(0, M_PI / 4))); // 10 different phases
model random_model(amplitude, phase)
```
This would run the tests with 10*10 randomly generated phases and amplitudes.

So I think this rng constructor should be left out.

joakim-hove replied:
> The test only uses the first constructor though

No - that is wrong; the setup in the loop:
```C++
 for (int iens = 0; iens < ens_size; iens++)
        ens.emplace_back(rng);
```

eivindjahren commented on libres/tests/analysis/ies/test_ies_enkf_main.cpp:
```diff
@@ -38,51 +35,142 @@ TEST_CASE("ies_enkf_linalg_extract_active", "[analysis]") {
+        matrix_type *W = ies_enkf_alloc_activeW(data);
+        for (int i = 0; i < ens_size - 1; i++) {
+            for (int j = 0; j < ens_size - 1; j++) {
+                int i0 = i + (i > 0);
```
Seems much better now :)

eivindjahren commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -330,88 +190,182 @@ void ies_enkf_updateA(
+    std::vector<double> costJ(ens_size);
+    double costf = 0.0;
+    for (int i = 0; i < ens_size; i++) {
+        costJ[i] = matrix_column_column_dot_product(W, i, W, i) +
```
If I am not mistaken, then the calculation of `costf` was only relevant for `tecplot`? I guess this can be removed now.

markusdregi replied:
See comments below ⬇️

geirev commented:
I would retain the costf computation as this is important information regarding the convergence of the ies that is also output to a log file.

Get Outlook for iOS<https://aka.ms/o0ukef>

[cid:norce_426cc920-64ab-4893-95d0-c29d5dab670c.png]

NORCE Norwegian Research Centre AS
norceresearch.no<https://www.norceresearch.no/>

________________________________
From: Eivind Jahren ***@***.***>
Sent: Wednesday, December 22, 2021 7:44:36 AM
To: equinor/ert ***@***.***>
Cc: Geir Evensen ***@***.***>; Mention ***@***.***>
Subject: Re: [equinor/ert] Ies testing (PR #2412)


CAUTION: This email originated from outside of the organization. Do not click links or open attachments unless you recognize the sender and know the content is safe.


@eivindjahren commented on this pull request.

________________________________

In libres/lib/analysis/ies/ies_enkf.cpp<https://github.com/equinor/ert/pull/2412#discussion_r773637376>:

>
-    ies_enkf_data_fclose_log(data);
+    std::vector<double> costJ(ens_size);
+    double costf = 0.0;
+    for (int i = 0; i < ens_size; i++) {
+        costJ[i] = matrix_column_column_dot_product(W, i, W, i) +


If I am not mistaken, then the calculation of costf was only relevant for tecplot? I guess this can be removed now.

—
Reply to this email directly, view it on GitHub<https://github.com/equinor/ert/pull/2412#pullrequestreview-838117686>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AFN4BGBDVKXFIHXVBIQP5JLUSFXVJANCNFSM5IQ7QFRQ>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

markusdregi commented:
Good comment @geirev! I suggest the following: For the scope of this PR we stick to the decision of dropping the tecplot functionality. Instead, we log the `costf` values computed via the ERT logger so that it is possible to get a hold of.

Later, we add support for visualising the cost function similarly to how one can plot the misfit function independently of which method was used for history matching. But that is far beyond the scope of this PR and its following friends...

joakim-hove commented:
> For the scope of this PR we stick to the decision of dropping the tecplot functionality. Instead, we log the costf values computed via the ERT logger so that it is possible to get a hold of.

I have done that.

joakim-hove commented:
OK - I have addressed nearly comments, and hope this can be considered for review now. The current test failures seem to be related to GitHub actions infrastructure.

eivindjahren commented on libres/tests/analysis/ies/test_ies_enkf_main.cpp:
```diff
@@ -38,51 +35,142 @@ TEST_CASE("ies_enkf_linalg_extract_active", "[analysis]") {
     // Test that `ies_enkf_linalg_extract_active` does nothing when all observations and realizations are active
-    matrix_type *E = matrix_alloc(state_size, ens_size);
-    ies_enkf_linalg_extract_active(data, E, stdout, false);
-    REQUIRE(matrix_equal(Ein, E));
+    {
```
```suggestion
Section("ies_enkf_linalg_extract_active does nothing when all are active"){
```

joakim-hove replied:
I would like to keep this test implementation as it is. The reason is that the `ies_enkf_data` structure maintains state throughout the entire test, and by introducing sections now for this particular test I fear there will be changes in behavior due to accumulated state. One of the main goals of the ongoing ies refactoring (not yet in this PR) is to reduce this state, and manage it a more transparent way.

dafeda replied:
From my understanding of how `catch2` works, it is safe to use `SECTION` in this case, as the test-framework allows us to clean-up before each section is run.
Why is this different?

joakim-hove replied:
The situation is opposite - the test holdes on to state which is mutated systematically through the test. 

It is certainly possible to rewrite to use sections - I suffered from a little review fatigue at the time.

eivindjahren replied:
I had a check and the required change in the test is minimal: 

```cpp
    SECTION("ies_enkf_linalg_extract_active does nothing when all observations and realizations are active"){
        matrix_type *E = ies_enkf_alloc_activeE(data);
        REQUIRE(matrix_equal(Ein, E));
        matrix_free(E);
    }

    SECTION("ies_enkf_linalg_extract_active can deactivate an ensemble") {
        bool_vector_iset(ens_mask, 1, false);
        ies_enkf_data_update_ens_mask(data, ens_mask);
        matrix_type *E = ies_enkf_alloc_activeE(data);
        REQUIRE(matrix_iget(E, 0, 0) == 1.0);
        REQUIRE(matrix_iget(E, 1, 0) == 2.0);
        REQUIRE(matrix_iget(E, 2, 0) == 3.0);
        REQUIRE(matrix_get_rows(E) == 3);
        REQUIRE(matrix_get_columns(E) == 1);
        matrix_free(E);
    }

    SECTION("ies_enkf_linalg_extract_active can deactivate an observation") {
        bool_vector_iset(obs_mask, 1, false);
        ies_enkf_update_obs_mask(data, obs_mask);
        matrix_type *E = ies_enkf_alloc_activeE(data);
        REQUIRE(matrix_iget(E, 0, 0) == 1.0);
        REQUIRE(matrix_iget(E, 1, 0) == 3.0);
        REQUIRE(matrix_get_rows(E) == 2);
        REQUIRE(matrix_get_columns(E) == 2);
        matrix_free(E);
    }
```

joakim-hove replied:
> the last assert needs to be changed to REQUIRE(matrix_get_columns(E) == 2);:

That is because of the state carrying I have argued about. Anyway - I'll update.

joakim-hove replied:
Done

eivindjahren commented on libres/lib/analysis/ies/ies_enkf_data.cpp:
```diff
@@ -221,45 +202,26 @@ void ies_enkf_data_augment_initialE(ies_enkf_data_type *data,
     if (!data->A0) {
         // We store the initial ensemble to use it in final update equation                     (Line 11)
-        bool dbg = ies_enkf_config_get_ies_debug(data->config);
         int m_state_size = std::min(matrix_get_rows(A) - 1, 50);
```
with `dbg` removed, `m_state_size` and `m_ens_size` is unused.

joakim-hove replied:
Ok

eivindjahren commented on libres/lib/analysis/ies/ies_enkf_data.cpp:
```diff
@@ -221,45 +202,26 @@ void ies_enkf_data_augment_initialE(ies_enkf_data_type *data,
-        bool dbg = ies_enkf_config_get_ies_debug(data->config);
+        // We initialize data->W which will store W for use in next iteration                    (Line 9)
+        int ens_size = bool_vector_size(data->ens_mask);
         int m_ens_size = std::min(ens_size - 1, 16);
```
with `dbg` removed, `m_ens_size` is unused.

joakim-hove replied:
Ok

eivindjahren reviewed (changes requested):
Getting into the details now, so we should be getting close :)

joakim-hove commented on libres/tests/analysis/ies/test_ies_enkf_main.cpp:
```diff
@@ -38,51 +35,142 @@ TEST_CASE("ies_enkf_linalg_extract_active", "[analysis]") {
+        matrix_free(E);
+    }
 
     // Test that `ies_enkf_linalg_extract_active` can deactivate an observation
     bool_vector_iset(obs_mask, 1, false);
     ies_enkf_update_obs_mask(data, obs_mask);
+    {
```
See comment above.

joakim-hove commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -330,88 +192,190 @@ void ies_enkf_updateA(
+    const ies_enkf_config_type *ies_config = ies_enkf_data_get_config(data);
 
-    int obs_size_msk =
+    int nrobs_msk =
```
OK

joakim-hove commented on libres/tests/res_util/test_matrix.cpp:
```diff
@@ -70,27 +70,30 @@ TEST_CASE("matrix_subtract_row_mean", "[res_util]") {
-SCENARIO("matrix_dgesvx") {
+SCENARIO("Solving a linear equation with dgesvx") {
     GIVEN("A diagonal matrix A") {
         std::size_t N = 4;
```
OK

joakim-hove commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -330,88 +192,190 @@ void ies_enkf_updateA(
-    int obs_size_msk =
+    int nrobs_msk =
         ies_enkf_data_get_obs_mask_size(data); // Total number of observations
+    int nrobs_inp = matrix_get_rows(
```
OK

joakim-hove commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -330,88 +192,190 @@ void ies_enkf_updateA(
+    int nrobs_inp = matrix_get_rows(
+        Yin); // Number of active observations input in current iteration
+
     int ens_size_msk =
```
OK

joakim-hove commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -330,88 +192,190 @@ void ies_enkf_updateA(
-                if (bool_vector_iget(ens_mask, iens)) {
-                    i++;
-                    matrix_iset_safe(E, m, i, matrix_iget(dataE, iobs, iens));
+    ies_inversion_type ies_inversion =
```
OK

joakim-hove commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -330,88 +192,190 @@ void ies_enkf_updateA(
+    ies_inversion_type ies_inversion =
+        ies_enkf_config_get_ies_inversion(ies_config);
+    double truncation = ies_enkf_config_get_truncation(ies_config);
+    int subspace_dimension =
```
OK

joakim-hove commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -420,68 +384,18 @@ void ies_enkf_linalg_compute_AA_projection(const matrix_type *A, matrix_type *Y,
     int m_ens_size = std::min(ens_size - 1, 16);
```
OK

eivindjahren replied:
This one seems still unresolved.

joakim-hove replied:
It is getting a bit difficult to navigate the diff here - but I think the issue as such has been resolved; I had just not pushed the "Resolved" button.

joakim-hove commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -490,7 +404,7 @@ void ies_enkf_linalg_extract_active_W(const ies_enkf_data_type *data,
+                        matrix_type *S) {
     int ens_size = matrix_get_columns(W0);
     int nrobs = matrix_get_rows(S);
     int m_ens_size = std::min(ens_size - 1, 16);
```
OK

joakim-hove commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -671,15 +536,14 @@ void ies_enkf_linalg_exact_inversion(matrix_type *W0, const int ies_inversion,
-                                    const matrix_type *W0, FILE *log_fp) {
+                                    const matrix_type *W0) {
     int ens_size_msk = ies_enkf_data_get_ens_mask_size(data);
     int ens_size = matrix_get_columns(W0);
```
OK

joakim-hove commented on libres/tests/analysis/ies/test_ies_enkf_main.cpp:
```diff
@@ -38,51 +35,142 @@ TEST_CASE("ies_enkf_linalg_extract_active", "[analysis]") {
-    REQUIRE(matrix_iget(E_ens_deactivate, 0, 0) == 1.0);
-    REQUIRE(matrix_iget(E_ens_deactivate, 1, 0) == 2.0);
-    REQUIRE(matrix_iget(E_ens_deactivate, 2, 0) == 3.0);
-    REQUIRE(matrix_iget(E_ens_deactivate, 0, 1) == 0.0);
-    REQUIRE(matrix_iget(E_ens_deactivate, 1, 1) == 0.0);
-    REQUIRE(matrix_iget(E_ens_deactivate, 2, 1) == 0.0);
+    {
```
See comment above.

joakim-hove commented on libres/tests/res_util/test_matrix.cpp:
```diff
@@ -70,27 +70,30 @@ TEST_CASE("matrix_subtract_row_mean", "[res_util]") {
 }
 
-SCENARIO("matrix_dgesvx") {
+SCENARIO("Solving a linear equation with dgesvx") {
```
OK - have reordered the commits.

eivindjahren commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -330,88 +190,176 @@ void ies_enkf_updateA(
+        Yin); // Number of active realizations in current iteration
+    int state_size = matrix_get_rows(A);
+
+    double truncation = ies_enkf_config_get_truncation(ies_config);
```
truncation is unsed here.

joakim-hove replied:
OK

eivindjahren commented on libres/tests/analysis/ies/test_ies_enkf_main.cpp:
```diff
@@ -38,51 +35,142 @@ TEST_CASE("ies_enkf_linalg_extract_active", "[analysis]") {
-    ies_enkf_config_set_ies_debug(ies_enkf_data_get_config(data), false);
     ies_enkf_config_set_ies_logfile(ies_enkf_data_get_config(data),
                                     "log_test_ies_enkf_linalg_extract_active");
     FILE *test_log = ies_enkf_data_open_log(data);
```
test_log is unsed now because of the refactoring

eivindjahren commented:
There has been a lot of comments, and I guess its a bit frustrating with all of that, and I apologize. I think we are quite close now, so I hope the remaining comments aren't too discouraging. I am generally "to the point" on github, and keep small talk on teams and slack, and I think that doesn't help with the experience. I find that is generally the case with us as a team, that small talk happens elsewhere. I find it a bit difficult to have a conversational tone on github, so I would encourage contacting us on video or on slack to have a chat, it might make for a happier life in general :) .

I find there is usually more comments in reviews just as someone has changed teams and then it quiets down as people get to know each other. Also, there has been more comments as it seems we are doing something quite relevant for the future.

joakim-hove commented:
> There has been a lot of comments, and I guess its a bit frustrating with all of that, and I apologize.

Thank you. Yes this review process has been a bit frustrating, let us all make an effort to make the next round more enjoyable.

joakim-hove commented:
```
Run mypy -p ert -p ert3 --config-file .mypy-strict.ini
ert3/config/_ensemble_config.py:47: error: unused "type: ignore" comment
ert3/algorithms/_sensitivity.py:109: error: unused "type: ignore" comment
Found 2 errors in 2 files (checked 42 source files)
Error: Process completed with exit code 1.
```

dafeda commented:
> ```
> Run mypy -p ert -p ert3 --config-file .mypy-strict.ini
> ert3/config/_ensemble_config.py:47: error: unused "type: ignore" comment
> ert3/algorithms/_sensitivity.py:109: error: unused "type: ignore" comment
> Found 2 errors in 2 files (checked 42 source files)
> Error: Process completed with exit code 1.
> ```

This is due to some change in `mypy`.
It's not yet decided how to fix this, but it'll probably be fixed tomorrow.

joakim-hove commented:
It is a bit challenging to keep track of all comments - but I think they are all addressed now and hope this can be merged.

oyvindeide commented on libres/lib/analysis/ies/ies_enkf.cpp:
```diff
@@ -330,88 +190,175 @@ void ies_enkf_updateA(
+    /* COMPUTE NEW ENSEMBLE SOLUTION FOR CURRENT ITERATION  Ei=A0*X (Line 11)*/
+    int m_ens_size = std::min(ens_size - 1, 16);
+    int m_state_size = std::min(state_size - 1, 3);
+    matrix_pretty_fprint_submat(A, "A^f", "%11.5f", log_fp, 0, m_state_size, 0,
```
This used to be behind a `dbg` flag, is it a leftover from debugging? Same a few lines below.

joakim-hove replied:
OK

dafeda commented on libres/tests/analysis/test_compare_initX.cpp:
```diff
@@ -0,0 +1,120 @@
+
+            matrix_type *X1 = matrix_alloc(ens_size, ens_size);
+            matrix_type *X2 = matrix_alloc(ens_size, ens_size);
+            std_enkf_set_double(std_enkf_data, ENKF_TRUNCATION_KEY_,
```
Any reason not to use `std_enkf_set_truncation` directly, instead of the wrapper `std_enkf_set_double`?

Also, I tried running the test with truncation set to `1.0` and to `0.01`:

`std_enkf_set_truncation(std_enkf_data, GENERATE(1.0, DEFAULT_ENKF_TRUNCATION_, 0.90, 0.01));`

The tests pass.
Does that seem reasonable to you?

joakim-hove replied:
> Any reason not to use std_enkf_set_truncation directly, instead of the wrapper std_enkf_set_double

Historical reason - closer to user-config facing implementation of `analysis_module_set_var()`. Have changed it.

joakim-hove replied:
>The tests pass.
Does that seem reasonable to you?

I guess you are alluding to the values 1.0 and 0.01? The value of 1.0 will mean that all singular values in the SVD are used - this will potentially become quite numerical unstable and prone to overfitting. The value of 0.01 means that only 1% of the variation will be included - essentially that will imply no singular values, and not very meaningful results. But - the current test just verifies that two different methods to initialize the X matrix give the same result, it does not in any way evaluate the quality of the results. So - yes I would expect this test to pass even if somewhat weird truncation values are used.

dafeda replied:
I was alluding to 1.0 and 0.01.
Perhaps we don't need to test with different truncations if they don't matter?

joakim-hove replied:
> Perhaps we don't need to test with different truncations if they don't matter?

Well - there is a point in testing with different truncations to verify that the truncation value can be modified for both `initX()` implementations.

dafeda commented:
> Thank you. Yes this review process has been a bit frustrating, let us all make an effort to make the next round more enjoyable.

Sorry to hear that you’ve found this review process frustrating.
We are interesting in hearing your thoughts on this, but that is perhaps better left to a video-chat.

For the record, I think this code review has been conducted in a professional and thorough manner.
Exactly how reviews should be in my opinion.

dafeda reviewed (approved):
LGTM 👍 

Looks like @pinkwah fixed the `mypy` issues and has merged to master, so tests should pass after rebase.
